api/proto/teleport/legacy/types/types.proto

```diff
@@ -4743,3 +4743,106 @@ enum RequireMFAType {
   // requiring double touch on session requests, normal Session MFA is disabled.
   HARDWARE_KEY_TOUCH = 3;
 }
+
+// Plugin describes a single instance of a Teleport Plugin
+message PluginV1 {
+  // Kind is the plugin resource kind.
+  string Kind = 1 [(gogoproto.jsontag) = "kind"];
+  // SubKind is an optional resource subkind.
+  string SubKind = 2 [(gogoproto.jsontag) = "sub_kind,omitempty"];
+  // Version is the resource version.
+  string Version = 3 [(gogoproto.jsontag) = "version"];
+
+  // Metadata is the resource metadata.
+  Metadata Metadata = 4 [
+    (gogoproto.nullable) = false,
+    (gogoproto.jsontag) = "metadata"
+  ];
+
+  PluginSpecV1 Spec = 5 [
+    (gogoproto.nullable) = false,
+    (gogoproto.jsontag) = "spec"
+  ];
+
+  PluginStatusV1 Status = 6 [
+    (gogoproto.nullable) = false,
+    (gogoproto.jsontag) = "status"
+  ];
+
+  // Credentials are "live" credentials to the 3rd party API.
+  // These are considered secrets.
+  PluginCredentialsV1 Credentials = 7 [(gogoproto.jsontag) = "credentials,omitempty"];
+}
+
+message PluginSpecV1 {
+  option (gogoproto.equal) = true;
+
+  // Settings contain provider-specific plugin options.
+  oneof Settings {
+    // Settings for the Slack access plugin
+    PluginSlackAccessSettings SlackAccessPlugin = 1 [(gogoproto.jsontag) = "slack_access,omitempty"];
+  }
+}
+
+message PluginSlackAccessSettings {
+  option (gogoproto.equal) = true;
+
+  string FallbackChannel = 1 [(gogoproto.jsontag) = "fallback_channel"];
+}
+
+message PluginInitialCredentialsV1 {
+  oneof Credentials {
+    PluginOAuth2AuthorizationCodeCredentials OAuth2AuthorizationCode = 1 [(gogoproto.jsontag) = "oauth2_authorization_code,omitempty"];
+  }
+}
+
+message PluginOAuth2AuthorizationCodeCredentials {
+  string AuthorizationCode = 1 [(gogoproto.jsontag) = "authorization_code"];
+  string RedirectURI = 2 [(gogoproto.jsontag) = "redirect_uri"];
+}
+
+// PluginStatus is the user-facing status for the plugin instance.
+message PluginStatusV1 {
+  PluginStatusCode Code = 1 [(gogoproto.jsontag) = "code"];
+}
+
+enum PluginStatusCode {
+  // UNKNOWN is the default value when the plugin has not reported its status yet.
+  UNKNOWN = 0;
+  // RUNNING means the plugin reports running successfully.
+  RUNNING = 1;
+  // OTHER_ERROR indicates that an otherwise-unspecified error has been encountered.
+  OTHER_ERROR = 2;
+  // UNAUTHORIZED indicates that plugin is not able to authenticate to the 3rd party API.
+  // This could be a result of e.g. the user revoking the authorization on the API provider's side.
+  UNAUTHORIZED = 3;
+
+  // SLACK_NOT_IN_CHANNEL is a Slack-specific status code that indicates
+  // that the bot has not been invited to a channel that it is configured to post in.
+  SLACK_NOT_IN_CHANNEL = 10;
+  reserved 11 to 19; // future Slack-specific codes
```

Comment:
I'd rather have this as a separate field in `PluginStatusV1`.

Reply:
Could you elaborate on what your suggestion would look like & what the benefits are?

Reply:
I was thinking of something like this, where the `Code` would be set to `OTHER_ERROR` or maybe `UNAUTHORIZED` depending on the plugin-specific error.
```proto
message PluginStatusV1 {
  PluginStatusCode Code = 1 [(gogoproto.jsontag) = "code"];
  oneof PluginSpecific {
    PluginSlackStatus Slack = 1;
  }
}

enum PluginSlackStatus {
  // The bot has not been invited to a channel that it is configured to post in.
  NOT_IN_CHANNEL = 1;
}
```

Thinking about it a bit more, this is probably worse than just having a single enum for Code 😬

Reply:
I don't think we should be reserving contiguous blocks of codes for the different plugins, however.